diff --git a/common/common.cpp b/common/common.cpp
--- a/llama.cpp/common/common.cpp
+++ b/llama.cpp/common/common.cpp
@@ -889,6 +889,18 @@ std::string fs_get_cache_directory() {
         cache_directory = std::getenv("HOME") + std::string("/Library/Caches/");
 #elif defined(_WIN32)
         cache_directory = std::getenv("LOCALAPPDATA");
+// llamafile START
+#elif defined(COSMOCC)
+        // We don't know what OS we are running on at compile time, just CPU architecture.
+        // try various environment variables, fall back to ~/.cache.
+         if (std::getenv("LOCALAPPDATA")) {
+             cache_directory = std::getenv("LOCALAPPDATA");
+        } else if (std::getenv("XDG_CACHE_HOME")) {
+             cache_directory = std::getenv("XDG_CACHE_HOME");
+         } else {
+             cache_directory = std::getenv("HOME") + std::string("/.cache/");
+         }
+// llamafile END
 #else
 #  error Unknown architecture
 #endif
